src/devices/nm-device.c

```diff
@@ -7881,51 +7900,42 @@ dhcp6_get_duid (NMDevice *self, NMConnection *connection, GBytes *hwaddr, NMDhcp
 
 		g_checksum_get_digest (sum, sha256_digest, &len);
 		g_checksum_free (sum);
-	}
 
-#define EPOCH_DATETIME_THREE_YEARS  (356 * 24 * 3600 * 3)
-	if (nm_streq0 (duid, "ll")) {
-		duid_out = generate_duid_ll (g_bytes_get_data (hwaddr, NULL));
+		if (nm_streq (duid, "stable-ll")) {
+			duid_out = generate_duid_ll (sha256_digest);
+		} if (nm_streq (duid, "stable-llt")) {
+			gint64 time;
 
-	} else if (nm_streq0 (duid, "llt")) {
-		gint64 time;
-
-		time = nm_utils_secret_key_get_timestamp ();
-		if (!time) {
-			duid_error = "cannot retrieve the secret key timestamp";
-			goto end;
-		}
-
-		duid_out = generate_duid_llt (g_bytes_get_data (hwaddr, NULL), time);
-	} else if (nm_streq0 (duid, "stable-ll")) {
-		duid_out = generate_duid_ll (sha256_digest);
+#define EPOCH_DATETIME_THREE_YEARS  (356 * 24 * 3600 * 3)
 
-	} else if (nm_streq0 (duid, "stable-llt")) {
-		gint64 time;
+			/* We want a variable time between the secret_key timestamp and three years
+			 * before. Let's compute the time (in seconds) from 0 to 3 years; then we'll
+			 * subtract it from the secret_key timestamp.
+			 */
+			time = nm_utils_secret_key_get_timestamp ();
+			if (!time) {
+				duid_error = "cannot retrieve the secret key timestamp";
+				goto out_fail;
+			}
+			/* don't use too old timestamps. They cannot be expressed in DUID-LLT and
+			 * would all be truncated to zero. */
+			time = NM_MAX (time, EPOCH_DATETIME_200001010000 + EPOCH_DATETIME_THREE_YEARS);
+			time -= (unaligned_read_be32 (&sha256_digest[ETH_ALEN]) % EPOCH_DATETIME_THREE_YEARS);
 
-		/* We want a variable time between the secret_key timestamp and three years
-		 * before. Let's compute the time (in seconds) from 0 to 3 years; then we'll
-		 * subtract it from the secret_key timestamp.
-		 */
-		time = nm_utils_secret_key_get_timestamp ();
-		if (!time) {
-			duid_error = "cannot retrieve the secret key timestamp";
-			goto end;
+			duid_out = generate_duid_llt (sha256_digest, time);
+		} else {
+			nm_assert (nm_streq (duid, "stable-uuid"));
+			duid_out = generate_duid_uuid (sha256_digest, len);
 		}
-		/* don't use too old timestamps. They cannot be expressed in DUID-LLT and
-		 * would all be truncated to zero. */
-		time = NM_MAX (time, EPOCH_DATETIME_200001010000 + EPOCH_DATETIME_THREE_YEARS);
-		time -= (unaligned_read_be32 (&sha256_digest[ETH_ALEN]) % EPOCH_DATETIME_THREE_YEARS);
-
-		duid_out = generate_duid_llt (sha256_digest, time);
 
-	} else if (nm_streq0 (duid, "stable-uuid")) {
-		duid_out = generate_duid_uuid (sha256_digest, len);
+		goto out_good;
 	}
 
-	duid_error = "generation failed";
-end:
-	if (!duid_out) {
+	g_return_val_if_reached (NULL);
```

Comment:
We want to log the error but provide the backup random DUID here: wouldn't be better have just an nm_assert here?